Modules/Core/Common/test/itkCopyGTest.cxx

```diff
@@ -0,0 +1,67 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         https://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+
+// First include the header file to be tested:
+#include "itkCopy.h"
+#include <gtest/gtest.h>
+
+#include <array>
+#include <vector>
+
+
+// Checks that an object returned by `itk::Copy(original)` compares equal to the original, even though it has a
+// different address.
+TEST(Copy, EqualToTheOriginalButHasDifferentAddress)
+{
+  const auto check = [](const auto & original) {
+    auto && copy = itk::Copy(original);
+
+    // The returned `copy` has an equal value...
+    EXPECT_EQ(copy, original);
+
+    // ... but a different address!
+    EXPECT_NE(&copy, &original);
+  };
+
+  check(42);
+  check(std::vector<std::intmax_t>{ 1, 2, 3, 4 });
+  check(std::array<std::intmax_t, 4>{ 1, 2, 3, 4 });
+}
+
+
+// Tests that `itk::Copy` allows storing a copy of a value in a variable declared by `auto`, without a warning.
+TEST(Copy, AllowsUsingAutoKeywordCopyingWithoutWarnings)
```

Comment:
@issakomi Can you please tell us if the proposed `itk::Copy` would indeed silence those Coverity AUTO_CAUSES_COPY warnings? Specifically, is this `AllowsUsingAutoKeywordCopyingWithoutWarnings` unit test indeed warning-free? And do those AUTO_CAUSES_COPY appear when the `itk::Copy` calls in there are removed?

Reply:
I will try. BTW, I don't run a Coverity scan for the entire ITK. I scan my applications from time to time and some headers from ITK are also scanned if they are enabled somehow. I'll try to reproduce it. IMHO we could just ignore the warnings.

Reply:
@issakomi For a specific application, I think those auto-causes-copy warnings can usually be ignored safely. But I think there is a policy for ITK to address warnings as much as possible. And it's also harder to ignore these specific warnings now that they come from at least three different tools (Intellisense/Linter, MSVC compiler and Coverity).

Reply:
@issakomi I just extended the `AllowsUsingAutoKeywordCopyingWithoutWarnings` to include structured bindings. Hope you can also try that out on Coverty. Specifically:

```cpp
  auto originalPair =
    std::make_pair(std::vector<std::intmax_t>{ 1, 2, 3, 4 }, std::array<std::intmax_t, 4>{ 1, 2, 3, 4 });
  const auto [first, second] = itk::Copy(originalPair);
```

Does the `itk::Copy` call here suppress another AUTO_CAUSES_COPY warning? Would the AUTO_CAUSES_COPY warning occur without the `itk::Copy` call (when simply doing `const auto [first, second] = originalPair`)?

I did already observe that the `itk::Copy` call in the context of structured bindings suppresses such warnings from Intellisense/Linter.

Reply:
> And it's also harder to ignore these specific warnings now that they come from at least three different tools

What I mean is that AUTO_CAUSES_COPY can be skipped if the object is small and trivial to copy. I have no experience with IntelliSense Code Linter. Coverity promises to suggest 'auto &' only if possible. The new _Copy_ function is a way to silence the warning, IMHO, it's the same thing as ignoring it.
BTW, maybe just create a test project for Coverity tests? Any opensource project can be scanned at https://scan.coverity.com. A full ITK scan will generate a ton of defects, mostly from third party modules.

Reply:
If necessary, please create a small test project for Coverity tests. Thank you.

Reply:
Thanks for doing those Coverity tests, @issakomi  In the mean time, I did further improve the unit tests. Especially by introducing a local function `GetReference` (just for the tests). Because I realized that those warnings appear specifically when there is a _reference_ at the right hand side (rhs). And that's exactly when `itk::Copy(rhs)` should be useful.

Thanks for suggesting my own Coverity tests, I think I'll do so later. But I think the MSVC warnings are quite similar, and I already have MSVC installed 😃.

Reply:
The reason for the suggestion of a special repository for _Coverity_ tests is that I just push a code somewhere in _main.cpp_ of my app and upload to _Coverity_ scan (scans are limited).

It is rather easy (github login can be used for _scan.coverity.com_), but a project have to be approved.

Short example on Linux:
 - download _Coverity_ archive, unpack in home directory
 - `export PATH=/home/r/cov-analysis-linux64-2023.6.2/bin:$PATH`
 - configure something like _mypoject/coverity-build_ and `cd mypoject/coverity-build`
 - `cov-build --dir cov-int make -j6`
 - `tar cf cov-int.tar cov-int`
 - `gzip cov-int.tar`
 - upload cov-int.tar.gz to scan portal (AFAIK, the name must always be _cov-int_)
 - wait for result of the scan

There are many option for automation, of course. Also a good idea may be to use "system" libraries  to reduce the number of defects (some parts of code can be skipped, but not really, the defects are not in the database, but seem to be still counted). There is also an option of "modeling" to suppress false-positives, but it seems to be a little over-complicated.

Probably commercial customers can do scan locally, not sure. Their portal might be a little better, IMHO, but the scan itself seems to be very good, not so nasty as many others and not so many false-positives (but there will be some of them, caution with silencing is a good idea, otherwise it is easy to make thing worse, as it happens very often with silencing of warnings from all the tools).